lang/cs/Org.Apache.REEF.Common/Telemetry/MetricsData.cs

```diff
@@ -0,0 +1,190 @@
+        /// <param name="me">The metric object returned.</param>
+        /// <returns>Boolean indicating if a metric object was succesfully retrieved.</returns>
+        public bool TryGetValue(string name, out IMetric me)
+        {
+            if (!_metricsMap.TryGetValue(name, out MetricTracker md))
+            {
+                me = null;
+                return false;
+            }
+            me = md.GetMetric();
+            return true;
+        }
+
+        /// <summary>
+        /// Gets all the registered metrics.
+        /// </summary>
+        /// <returns>IEnumerable of MetricData.</returns>
+        public IEnumerable<MetricTracker> GetMetrics()
+        {
+            return _metricsMap.Values;
+        }
+
+        /// <summary>
+        /// Updates metrics given another <see cref="MetricsData"/> object.
+        /// For every metric in the new set, if it is registered then update the value,
+        /// if it is not then add it to the registration.
+        /// </summary>
+        /// <param name="metrics">New metric values to be updated.</param>
+        internal void Update(MetricsData metrics)
+        {
+            foreach (var metric in metrics.GetMetrics())
+            {
+                _metricsMap.AddOrUpdate(metric.GetMetric().Name, metric, (k, v) => v.UpdateMetric(metric));
+            }
+        }
+
+        /// <summary>
+        /// Reset changed since last sink for each metric
+        /// </summary>
+        internal void Reset()
+        {
+            lock (_metricLock)
+            {
+                foreach (var tracker in _metricsMap.Values)
+                {
+                    tracker.ResetChangesSinceLastSink();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Convert the metric data to a collection of IMetric for sinking.
+        /// </summary>
+        /// <returns>A collection of metric records.</returns>
+        internal IEnumerable<KeyValuePair<string, MetricTracker.MetricRecord>> GetMetricsHistory()
+        {
+            var records = new List<KeyValuePair<string, MetricTracker.MetricRecord>>();
+            foreach (var me in _metricsMap)
+            {
+                var name = me.Key;  // name of metric
+                var data = me.Value;  // metric tracker
+                foreach (var record in data.GetMetricRecords())
+                {
+                    records.Add(new KeyValuePair<string, MetricTracker.MetricRecord>(name, record));
+                }
+            }
+            return records;
+        }
+
+        /// <summary>
+        /// The condition that triggers the sink. The condition can be modified later.
+        /// </summary>
+        /// <returns></returns>
+        internal bool TriggerSink(int metricSinkThreshold)
+        {
+            return _metricsMap.Values.Sum(e => e.ChangesSinceLastSink) > metricSinkThreshold;
+        }
+
+        public string Serialize()
+        {
+            lock (_metricLock)
```

Comment:
The _metricLock usage is inconsistent. We'll want to synchronize a lot of this logic. I'll take with you offline.